man/tcamgr.1
.TH "TCATEST" 3 "2009-06-21" "Man Page" "Tokyo Cabinet"

.SH NAME
tcamgr \- the command line utility of the abstract database API

.SH DESCRIPTION
.PP
The command `\fBtcamgr\fR' is a utility for test and debugging of the abstract database API and its applications.  `\fIname\fR' specifies the name of a database.  `\fIkey\fR' specifies the key of a record.  `\fIvalue\fR' specifies the value of a record.  `\fIparams\fR' specifies the tuning parameters.  `\fIfunc\fR' specifies the name of a function.  `\fIarg\fR' specifies the arguments of the function.  `\fIdest\fR' specifies the path of the destination file.
.PP
.RS
.br
\fBtcamgr create \fIname\fB\fR
.RS
Create a database file.
.RE
.br
\fBtcamgr inform \fIname\fB\fR
.RS
Print miscellaneous information to the standard output.
.RE
.br
\fBtcamgr put \fR[\fB\-sx\fR]\fB \fR[\fB\-sep \fIchr\fB\fR]\fB \fR[\fB\-dk\fR|\fB\-dc\fR|\fB\-dai\fR|\fB\-dad\fR]\fB \fIname\fB \fIkey\fB \fIvalue\fB\fR
.RS
Store a record.
.RE
.br
\fBtcamgr out \fR[\fB\-sx\fR]\fB \fR[\fB\-sep \fIchr\fB\fR]\fB \fIname\fB \fIkey\fB\fR
.RS
Remove a record.
.RE
.br
\fBtcamgr get \fR[\fB\-sx\fR]\fB \fR[\fB\-sep \fIchr\fB\fR]\fB \fR[\fB\-px\fR]\fB \fR[\fB\-pz\fR]\fB \fIname\fB \fIkey\fB\fR
.RS
Print the value of a record.
.RE
.br
\fBtcamgr list \fR[\fB\-sep \fIchr\fB\fR]\fB \fR[\fB\-m \fInum\fB\fR]\fB \fR[\fB\-pv\fR]\fB \fR[\fB\-px\fR]\fB \fR[\fB\-fm \fIstr\fB\fR]\fB \fIname\fB\fR
.RS
Print keys of all records, separated by line feeds.
.RE
.br
\fBtcamgr optimize \fIname\fB \fIparams\fB\fR
.RS
Optimize a database file.
.RE
.br
\fBtcamgr misc \fR[\fB\-sx\fR]\fB \fR[\fB\-sep \fIchr\fB\fR]\fB \fR[\fB\-px\fR]\fB \fIname\fB \fIfunc\fB \fR[\fB\fIarg\fB...\fR]\fB\fR
.RS
Call a versatile function for miscellaneous operations.
.RE
.br
\fBtcamgr map \fR[\fB\-fm \fIstr\fB\fR]\fB \fIname\fB \fIdest\fB\fR
.RS
Map records into another B+ tree database.
.RE
.br
\fBtcamgr version\fR
.RS
Print the version information of Tokyo Cabinet.
.RE
.RE
.PP
Options feature the following.
.PP
.RS
\fB\-sx\fR : the input data is evaluated as a hexadecimal data string.
.br
\fB\-sep \fIchr\fR\fR : specify the separator of the input data.
.br
\fB\-dk\fR : use the function `tcadbputkeep' instead of `tcadbput'.
.br
\fB\-dc\fR : use the function `tcadbputcat' instead of `tcadbput'.
.br
\fB\-dai\fR : use the function `tcadbaddint' instead of `tcadbput'.
.br
\fB\-dad\fR : use the function `tcadbadddouble' instead of `tcadbput'.
.br
\fB\-px\fR : the output data is converted into a hexadecimal data string.
.br
\fB\-pz\fR : do not append line feed at the end of the output.
.br
\fB\-m \fInum\fR\fR : specify the maximum number of the output.
.br
\fB\-pv\fR : print values of records also.
.br
\fB\-fm \fIstr\fR\fR : specify the prefix of keys.
.br
.RE
.PP
This command returns 0 on success, another on failure.

.SH SEE ALSO
.PP
.BR tcatest (1),
.BR tcadb (3),
.BR tokyocabinet (3)


ChangeLog
2009-07-03  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tctdb.c (tctdbidxgetbyftsunion): efficiency couting sort was improved.

	- Release: 1.4.29

2009-06-22  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tcutil.c (tcstrutfnorm, tcstrkwic, tcstrtokenize): new functions.

	* tcutil.c (tcstrutfnorm, tcstrucsnorm): "TCUNWIDTH" option was added.

	* tctdb.c (tctdbsetinvcache, tctdbidxsyncicc, tctdbidxcmpkey): new functions.

	* tctdb.c (tctdbmetasearch, tctdbget4, tctdbqrykwic): new functions.

	* tctdb.c (tctdbsetindeximpl): the xmsiz parameter is now configured.

	- Release: 1.4.28

2009-06-16  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tcutil.c (tcstrskipspc, tcstrucsnorm): new functions.

	* tctdb.c (tctdbqryaddcond): full-text search operators were added.

	* tctdb.c (tctdbsetindex): q-gram inverted index was added.

	* tctdb.c (tctdbidxputqgram, tctdbidxoutqgram, tctdbidxgetbyfts): new functions.

	- Release: 1.4.27

2009-06-13  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tctdb.c (tctdbsetindex, tctdbsearchimpl): token inverted index was added.

	* tctdb.c (tctdbidxputone, tctdbidxoutone): new functions.

	* tctdb.c (tctdbidxputtoken, tctdbidxouttoken, tctdbidxgetbytokens): new functions.

	* tcadb.c (tcadbput, tcadbputkeep, tcadbputcat): capnum behaviour was modified.

	- Release: 1.4.26

2009-06-09  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tcutil.c (tccstrescape, tccstrunescape, tcjsonescape, tcjsonunescape): new functions.

	* tcutil.c (tcpathlock, tcpathunlock): new functions.

	* tchdb.c (tchdbopen): double opening detection was added.

	* tchdb.c (tchdbtranbegin): a bug of invalid flag setting was fixed.

	* tcfdb.c (tcfdbopen): double opening detection was added.

	* tcfdb.c (tcfdbtranbegin): a bug of invalid flag setting was fixed.

	* tctdb.c (tctdbiternext3): new function.

	- Release: 1.4.25

2009-06-04  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tcutil.c (tcwwwformencode, tcwwwformdecode): new functions.

	* tcutil.c (tctmplnew, tctmpldel, tctmplload, tctmpldump): new functions.

	* tcutil.c (tcmapget4, tctreeget4): new functions.

	* tcutil.c (tclistprintf, tcmapprintf, tctreeprintf): new functions.

	* tcucodec.c (runtmpl, proctmpl): new functions.

	- Release: 1.4.24

2009-05-24  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tcutil.c (tcmapiterinit2, tcmdbiterinit2): new functions.

	* tchdb.c (tchdbiterinit2, tchdbiterjumpimpl): new functions.

	* tcfdb.c (tcfdbiterinit2, tcfdbiterjumpimpl): new functions.

	* tctdb.c (tctdbiterinit2): new function.

	- Release: 1.4.23

2009-05-21  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tcadb.c (tcadbmisc): "sync", "optimize", "vanish" commands were added.

	- Release: 1.4.22

2009-05-10  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tchdb.c (tchdbsetdfunit, tchdbdfunit, tchdbdefrag): new functions.

	* tchdb.c (tchdbdefragimpl, tchdbfbptrim, tchdbshiftrec): new functions.

	* tcbdb.c (tcbdbsetdfunit, tcbdbdfunit, tcbdbdefrag): new functions.

	* tctdb.c (tctdbsetdfunit, tctdbdfunit, tctdbdefrag): new functions.

	* tcadb.c (tcadbopen): "dfunit" parameter was added.

	* tcadb.c (tcadbmisc): "defrag" command was added.

	- Release: 1.4.21

2009-05-08  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tcutil.h: signedness of some members of TCMAP and TCTREE were changed.

	* tcutil.c (tchexdecode): a bug related to handling space characters was fixed.

	- Release: 1.4.20

2009-04-30  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tctdb.c (tctdbidxhash, tctdbqryproc2, tctdbqrysearchout2): new functions.

	* tctdb.c (tctdbsearchimpl, tctdbidxput, tctdbidxout): the index format was modified.

	* tcadb.c (tcadbmisc): "out" option of "search" command became non-atomic.

	- Release: 1.4.19

2009-04-25  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tcutil.c (tcatoih): new function.

	* tcbdb.c (tcbdbputimpl): cache adjustment on the putkeep mode was added.

	* tcadb.c (tcadbsetskel): new function.

	- Release: 1.4.18

2009-04-23  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tcutil.c (tcmapput, tcmapout, tcmapget): memory usage was reduced.

	* tcutil.c (tcsysinfo): new function.

	* tcutil.c (tcatoix): "strtold" was replaced by own implementation.

	* tctdb.c (tctdbqryidxfetch): new function.

	- Release: 1.4.17

2009-04-20  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tcbdb.c (tcbdbcuroutimpl): shift mechanism of cursors on deleted leaves was added.

	* tcbdb.c (tcbdbleafcheck): new function.

	- Release: 1.4.16

2009-04-08  Mikio Hirabayashi  <mikio@users.sourceforge.net>

	* tcutil.c (tcsleep): new function.

	* tchdb.c (tchdbtranbegin): locking algorithm was modified.

	* tchdb.c (tchdblocktran, tchdbunlocktran): abolished.

	* tcbdb.c (tcbdbtranbegin): locking algorithm was modified.

	* tcbdb.c (tcbdbcurjumpimpl): a bug related to cursor initialization was fixed.

	* tcfdb.c (tcbdbtranbegin): locking algorithm was modified.

	* tcfdb.c (tcfdblocktran, tcfdbunlocktran): abolished.

	* tctdb.c (tctdbqryallcondmatch): a bug related to handling null value was fixed.

	* tcadb.c (tcadboptimize, tcadbpath): new functions.

	- Release: 1.4.15


tcamgr.c
/*************************************************************************************************
 * The command line utility of the abstract database API
 *                                                      Copyright (C) 2006-2009 Mikio Hirabayashi
 * This file is part of Tokyo Cabinet.
 * Tokyo Cabinet is free software; you can redistribute it and/or modify it under the terms of
 * the GNU Lesser General Public License as published by the Free Software Foundation; either
 * version 2.1 of the License or any later version.  Tokyo Cabinet is distributed in the hope
 * that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 * You should have received a copy of the GNU Lesser General Public License along with Tokyo
 * Cabinet; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307 USA.
 *************************************************************************************************/


#include <tcutil.h>
#include <tcbdb.h>
#include <tcadb.h>
#include "myconf.h"


/* global variables */
const char *g_progname;                  // program name


/* function prototypes */
int main(int argc, char **argv);
static void usage(void);
static void printerr(TCADB *adb);
static int sepstrtochr(const char *str);
static char *strtozsv(const char *str, int sep, int *sp);
static int printdata(const char *ptr, int size, bool px, int sep);
static bool mapbdbproc(void *map, const char *kbuf, int ksiz, const char *vbuf, int vsiz,
                       void *op);
static int runcreate(int argc, char **argv);
static int runinform(int argc, char **argv);
static int runput(int argc, char **argv);
static int runout(int argc, char **argv);
static int runget(int argc, char **argv);
static int runlist(int argc, char **argv);
static int runoptimize(int argc, char **argv);
static int runmisc(int argc, char **argv);
static int runmap(int argc, char **argv);
static int runversion(int argc, char **argv);
static int proccreate(const char *name);
static int procinform(const char *name);
static int procput(const char *name, const char *kbuf, int ksiz, const char *vbuf, int vsiz,
                   int dmode);
static int procout(const char *name, const char *kbuf, int ksiz);
static int procget(const char *name, const char *kbuf, int ksiz, int sep, bool px, bool pz);
static int proclist(const char *name, int sep, int max, bool pv, bool px, const char *fmstr);
static int procoptimize(const char *name, const char *params);
static int procmisc(const char *name, const char *func, const TCLIST *args, int sep, bool px);
static int procmap(const char *name, const char *dest, const char *fmstr);
static int procversion(void);


/* main routine */
int main(int argc, char **argv){
  g_progname = argv[0];
  if(argc < 2) usage();
  int rv = 0;
  if(!strcmp(argv[1], "create")){
    rv = runcreate(argc, argv);
  } else if(!strcmp(argv[1], "inform")){
    rv = runinform(argc, argv);
  } else if(!strcmp(argv[1], "put")){
    rv = runput(argc, argv);
  } else if(!strcmp(argv[1], "out")){
    rv = runout(argc, argv);
  } else if(!strcmp(argv[1], "get")){
    rv = runget(argc, argv);
  } else if(!strcmp(argv[1], "list")){
    rv = runlist(argc, argv);
  } else if(!strcmp(argv[1], "optimize")){
    rv = runoptimize(argc, argv);
  } else if(!strcmp(argv[1], "misc")){
    rv = runmisc(argc, argv);
  } else if(!strcmp(argv[1], "map")){
    rv = runmap(argc, argv);
  } else if(!strcmp(argv[1], "version") || !strcmp(argv[1], "--version")){
    rv = runversion(argc, argv);
  } else {
    usage();
  }
  return rv;
}


/* print the usage and exit */
static void usage(void){
  fprintf(stderr, "%s: the command line utility of the abstract database API\n", g_progname);
  fprintf(stderr, "\n");
  fprintf(stderr, "usage:\n");
  fprintf(stderr, "  %s create name\n", g_progname);
  fprintf(stderr, "  %s inform name\n", g_progname);
  fprintf(stderr, "  %s put [-sx] [-sep chr] [-dk|-dc|-dai|-dad] name key value\n", g_progname);
  fprintf(stderr, "  %s out [-sx] [-sep chr] name key\n", g_progname);
  fprintf(stderr, "  %s get [-sx] [-sep chr] [-px] [-pz] name key\n", g_progname);
  fprintf(stderr, "  %s list [-sep chr] [-m num] [-pv] [-px] [-fm str] name\n", g_progname);
  fprintf(stderr, "  %s optimize name [params]\n", g_progname);
  fprintf(stderr, "  %s misc [-sx] [-sep chr] [-px] name func [arg...]\n", g_progname);
  fprintf(stderr, "  %s map [-fm str] name dest\n", g_progname);
  fprintf(stderr, "  %s version\n", g_progname);
  fprintf(stderr, "\n");
  exit(1);
}


/* get the character of separation string */
static int sepstrtochr(const char *str){
  if(!strcmp(str, "\\t")) return '\t';
  if(!strcmp(str, "\\r")) return '\r';
  if(!strcmp(str, "\\n")) return '\n';
  return *(unsigned char *)str;
}


/* encode a string as a zero separaterd string */
static char *strtozsv(const char *str, int sep, int *sp){
 int size = strlen(str);
 char *buf = tcmemdup(str, size);
 for(int i = 0; i < size; i++){
   if(buf[i] == sep) buf[i] = '\0';
 }
 *sp = size;
 return buf;
}


/* print error information */
static void printerr(TCADB *adb){
  const char *path = tcadbpath(adb);
  fprintf(stderr, "%s: %s: error\n", g_progname, path ? path : "-");
}


/* print record data */
static int printdata(const char *ptr, int size, bool px, int sep){
  int len = 0;
  while(size-- > 0){
    if(px){
      if(len > 0) putchar(' ');
      len += printf("%02X", *(unsigned char *)ptr);
    } else if(sep > 0){
      if(*ptr == '\0'){
        putchar(sep);
      } else {
        putchar(*ptr);
      }
      len++;
    } else {
      putchar(*ptr);
      len++;
    }
    ptr++;
  }
  return len;
}


/* mapping function */
static bool mapbdbproc(void *map, const char *kbuf, int ksiz, const char *vbuf, int vsiz,
                       void *op){
  bool err = false;
  if(!tcadbmapbdbemit(map, kbuf, ksiz, vbuf, vsiz)) err = true;
  return !err;
}


/* parse arguments of create command */
static int runcreate(int argc, char **argv){
  char *name = NULL;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      usage();
    } else if(!name){
      name = argv[i];
    } else {
      usage();
    }
  }
  if(!name) usage();
  int rv = proccreate(name);
  return rv;
}


/* parse arguments of inform command */
static int runinform(int argc, char **argv){
  char *name = NULL;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      usage();
    } else if(!name){
      name = argv[i];
    } else {
      usage();
    }
  }
  if(!name) usage();
  name = tcsprintf("%s#mode=r", name);
  int rv = procinform(name);
  tcfree(name);
  return rv;
}


/* parse arguments of put command */
static int runput(int argc, char **argv){
  char *name = NULL;
  char *key = NULL;
  char *value = NULL;
  int dmode = 0;
  bool sx = false;
  int sep = -1;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-dk")){
        dmode = -1;
      } else if(!strcmp(argv[i], "-dc")){
        dmode = 1;
      } else if(!strcmp(argv[i], "-dai")){
        dmode = 10;
      } else if(!strcmp(argv[i], "-dad")){
        dmode = 11;
      } else if(!strcmp(argv[i], "-sx")){
        sx = true;
      } else if(!strcmp(argv[i], "-sep")){
        if(++i >= argc) usage();
        sep = sepstrtochr(argv[i]);
      } else {
        usage();
      }
    } else if(!name){
      name = argv[i];
    } else if(!key){
      key = argv[i];
    } else if(!value){
      value = argv[i];
    } else {
      usage();
    }
  }
  if(!name || !key || !value) usage();
  char *kbuf, *vbuf;
  int ksiz, vsiz;
  if(sx){
    kbuf = tchexdecode(key, &ksiz);
    vbuf = tchexdecode(value, &vsiz);
  } else if(sep > 0){
    kbuf = strtozsv(key, sep, &ksiz);
    vbuf = strtozsv(value, sep, &vsiz);
  } else {
    ksiz = strlen(key);
    kbuf = tcmemdup(key, ksiz);
    vsiz = strlen(value);
    vbuf = tcmemdup(value, vsiz);
  }
  int rv = procput(name, kbuf, ksiz, vbuf, vsiz, dmode);
  tcfree(vbuf);
  tcfree(kbuf);
  return rv;
}


/* parse arguments of out command */
static int runout(int argc, char **argv){
  char *name = NULL;
  char *key = NULL;
  bool sx = false;
  int sep = -1;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-sx")){
        sx = true;
      } else if(!strcmp(argv[i], "-sep")){
        if(++i >= argc) usage();
        sep = sepstrtochr(argv[i]);
      } else {
        usage();
      }
    } else if(!name){
      name = argv[i];
    } else if(!key){
      key = argv[i];
    } else {
      usage();
    }
  }
  if(!name || !key) usage();
  int ksiz;
  char *kbuf;
  if(sx){
    kbuf = tchexdecode(key, &ksiz);
  } else if(sep > 0){
    kbuf = strtozsv(key, sep, &ksiz);
  } else {
    ksiz = strlen(key);
    kbuf = tcmemdup(key, ksiz);
  }
  int rv = procout(name, kbuf, ksiz);
  tcfree(kbuf);
  return rv;
}


/* parse arguments of get command */
static int runget(int argc, char **argv){
  char *name = NULL;
  char *key = NULL;
  bool sx = false;
  int sep = -1;
  bool px = false;
  bool pz = false;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-sx")){
        sx = true;
      } else if(!strcmp(argv[i], "-sep")){
        if(++i >= argc) usage();
        sep = sepstrtochr(argv[i]);
      } else if(!strcmp(argv[i], "-px")){
        px = true;
      } else if(!strcmp(argv[i], "-pz")){
        pz = true;
      } else {
        usage();
      }
    } else if(!name){
      name = argv[i];
    } else if(!key){
      key = argv[i];
    } else {
      usage();
    }
  }
  if(!name || !key) usage();
  int ksiz;
  char *kbuf;
  if(sx){
    kbuf = tchexdecode(key, &ksiz);
  } else if(sep > 0){
    kbuf = strtozsv(key, sep, &ksiz);
  } else {
    ksiz = strlen(key);
    kbuf = tcmemdup(key, ksiz);
  }
  name = tcsprintf("%s#mode=r", name);
  int rv = procget(name, kbuf, ksiz, sep, px, pz);
  tcfree(name);
  tcfree(kbuf);
  return rv;
}


/* parse arguments of list command */
static int runlist(int argc, char **argv){
  char *name = NULL;
  int sep = -1;
  int max = -1;
  bool pv = false;
  bool px = false;
  char *fmstr = NULL;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-sep")){
        if(++i >= argc) usage();
        sep = sepstrtochr(argv[i]);
      } else if(!strcmp(argv[i], "-m")){
        if(++i >= argc) usage();
        max = tcatoix(argv[i]);
      } else if(!strcmp(argv[i], "-pv")){
        pv = true;
      } else if(!strcmp(argv[i], "-px")){
        px = true;
      } else if(!strcmp(argv[i], "-fm")){
        if(++i >= argc) usage();
        fmstr = argv[i];
      } else {
        usage();
      }
    } else if(!name){
      name = argv[i];
    } else {
      usage();
    }
  }
  if(!name) usage();
  name = tcsprintf("%s#mode=r", name);
  int rv = proclist(name, sep, max, pv, px, fmstr);
  tcfree(name);
  return rv;
}


/* parse arguments of optimize command */
static int runoptimize(int argc, char **argv){
  char *name = NULL;
  char *params = NULL;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      usage();
    } else if(!name){
      name = argv[i];
    } else if(!params){
      params = argv[i];
    } else {
      usage();
    }
  }
  if(!name) usage();
  int rv = procoptimize(name, params);
  return rv;
}


/* parse arguments of misc command */
static int runmisc(int argc, char **argv){
  char *name = NULL;
  char *func = NULL;
  TCLIST *args = tcmpoollistnew(tcmpoolglobal());
  bool sx = false;
  int sep = -1;
  bool px = false;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-sx")){
        sx = true;
      } else if(!strcmp(argv[i], "-sep")){
        if(++i >= argc) usage();
        sep = sepstrtochr(argv[i]);
      } else if(!strcmp(argv[i], "-px")){
        px = true;
      } else {
        usage();
      }
    } else if(!name){
      name = argv[i];
    } else if(!func){
      func = argv[i];
    } else {
      if(sx){
        int size;
        char *buf = tchexdecode(argv[i], &size);
        tclistpush(args, buf, size);
        tcfree(buf);
      } else if(sep > 0){
        int size;
        char *buf = strtozsv(argv[i], sep, &size);
        tclistpush(args, buf, size);
        tcfree(buf);
      } else {
        tclistpush2(args, argv[i]);
      }
    }
  }
  if(!name || !func) usage();
  int rv = procmisc(name, func, args, sep, px);
  return rv;
}


/* parse arguments of map command */
static int runmap(int argc, char **argv){
  char *name = NULL;
  char *dest = NULL;
  char *fmstr = NULL;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-fm")){
        if(++i >= argc) usage();
        fmstr = argv[i];
      } else {
        usage();
      }
    } else if(!name){
      name = argv[i];
    } else if(!dest){
      dest = argv[i];
    } else {
      usage();
    }
  }
  if(!name || !dest) usage();
  name = tcsprintf("%s#mode=r", name);
  int rv = procmap(name, dest, fmstr);
  tcfree(name);
  return rv;
}


/* parse arguments of version command */
static int runversion(int argc, char **argv){
  int rv = procversion();
  return rv;
}


/* perform create command */
static int proccreate(const char *name){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  if(!tcadbclose(adb)){
    printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform inform command */
static int procinform(const char *name){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  const char *path = tcadbpath(adb);
  if(!path) path = "(unknown)";
  printf("path: %s\n", path);
  const char *type = "(unknown)";
  switch(tcadbomode(adb)){
  case ADBOVOID: type = "not opened"; break;
  case ADBOMDB: type = "on-memory hash database"; break;
  case ADBONDB: type = "on-memory tree database"; break;
  case ADBOHDB: type = "hash database"; break;
  case ADBOBDB: type = "B+ tree database"; break;
  case ADBOFDB: type = "fixed-length database"; break;
  case ADBOTDB: type = "table database"; break;
  }
  printf("database type: %s\n", type);
  printf("record number: %llu\n", (unsigned long long)tcadbrnum(adb));
  printf("size: %llu\n", (unsigned long long)tcadbsize(adb));
  if(!tcadbclose(adb)){
    printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform put command */
static int procput(const char *name, const char *kbuf, int ksiz, const char *vbuf, int vsiz,
                   int dmode){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  switch(dmode){
  case -1:
    if(!tcadbputkeep(adb, kbuf, ksiz, vbuf, vsiz)){
      printerr(adb);
      err = true;
    }
    break;
  case 1:
    if(!tcadbputcat(adb, kbuf, ksiz, vbuf, vsiz)){
      printerr(adb);
      err = true;
    }
    break;
  case 10:
    if(tcadbaddint(adb, kbuf, ksiz, tcatoi(vbuf)) == INT_MIN){
      printerr(adb);
      err = true;
    }
    break;
  case 11:
    if(isnan(tcadbadddouble(adb, kbuf, ksiz, tcatof(vbuf)))){
      printerr(adb);
      err = true;
    }
    break;
  default:
    if(!tcadbput(adb, kbuf, ksiz, vbuf, vsiz)){
      printerr(adb);
      err = true;
    }
    break;
  }
  if(!tcadbclose(adb)){
    if(!err) printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform out command */
static int procout(const char *name, const char *kbuf, int ksiz){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  if(!tcadbout(adb, kbuf, ksiz)){
    printerr(adb);
    err = true;
  }
  if(!tcadbclose(adb)){
    if(!err) printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform get command */
static int procget(const char *name, const char *kbuf, int ksiz, int sep, bool px, bool pz){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  int vsiz;
  char *vbuf = tcadbget(adb, kbuf, ksiz, &vsiz);
  if(vbuf){
    printdata(vbuf, vsiz, px, sep);
    if(!pz) putchar('\n');
    tcfree(vbuf);
  } else {
    printerr(adb);
    err = true;
  }
  if(!tcadbclose(adb)){
    if(!err) printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform list command */
static int proclist(const char *name, int sep, int max, bool pv, bool px, const char *fmstr){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  if(fmstr){
    TCLIST *keys = tcadbfwmkeys2(adb, fmstr, max);
    for(int i = 0; i < tclistnum(keys); i++){
      int ksiz;
      const char *kbuf = tclistval(keys, i, &ksiz);
      printdata(kbuf, ksiz, px, sep);
      if(pv){
        int vsiz;
        char *vbuf = tcadbget(adb, kbuf, ksiz, &vsiz);
        if(vbuf){
          putchar('\t');
          printdata(vbuf, vsiz, px, sep);
          tcfree(vbuf);
        }
      }
      putchar('\n');
    }
    tclistdel(keys);
  } else {
    if(!tcadbiterinit(adb)){
      printerr(adb);
      err = true;
    }
    int ksiz;
    char *kbuf;
    int cnt = 0;
    while((kbuf = tcadbiternext(adb, &ksiz)) != NULL){
      printdata(kbuf, ksiz, px, sep);
      if(pv){
        int vsiz;
        char *vbuf = tcadbget(adb, kbuf, ksiz, &vsiz);
        if(vbuf){
          putchar('\t');
          printdata(vbuf, vsiz, px, sep);
          tcfree(vbuf);
        }
      }
      putchar('\n');
      tcfree(kbuf);
      if(max >= 0 && ++cnt >= max) break;
    }
  }
  if(!tcadbclose(adb)){
    if(!err) printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform optimize command */
static int procoptimize(const char *name, const char *params){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  if(!tcadboptimize(adb, params)){
    printerr(adb);
    err = true;
  }
  if(!tcadbclose(adb)){
    if(!err) printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform misc command */
static int procmisc(const char *name, const char *func, const TCLIST *args, int sep, bool px){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  TCLIST *res = tcadbmisc(adb, func, args);
  if(res){
    for(int i = 0; i < tclistnum(res); i++){
      int rsiz;
      const char *rbuf = tclistval(res, i, &rsiz);
      printdata(rbuf, rsiz, px, sep);
      printf("\n");
    }
    tclistdel(res);
  } else {
    printerr(adb);
    err = true;
  }
  if(!tcadbclose(adb)){
    if(!err) printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform map command */
static int procmap(const char *name, const char *dest, const char *fmstr){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  TCBDB *bdb = tcbdbnew();
  if(!tcbdbopen(bdb, dest, BDBOWRITER | BDBOCREAT | BDBOTRUNC)){
    printerr(adb);
    tcbdbdel(bdb);
    tcadbdel(adb);
    return 1;
  }
  if(fmstr){
    TCLIST *keys = tcadbfwmkeys2(adb, fmstr, -1);
    if(!tcadbmapbdb(adb, keys, bdb, mapbdbproc, NULL, -1)){
      printerr(adb);
      err = true;
    }
    tclistdel(keys);
  } else {
    if(!tcadbmapbdb(adb, NULL, bdb, mapbdbproc, NULL, -1)){
      printerr(adb);
      err = true;
    }
  }
  if(!tcbdbclose(bdb)){
    printerr(adb);
    err = true;
  }
  tcbdbdel(bdb);
  if(!tcadbclose(adb)){
    printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform version command */
static int procversion(void){
  printf("Tokyo Cabinet version %s (%d:%s) for %s\n",
         tcversion, _TC_LIBVER, _TC_FORMATVER, TCSYSNAME);
  printf("Copyright (C) 2006-2009 Mikio Hirabayashi\n");
  return 0;
}



// END OF FILE
